specify that types can be omitted from augmenting declarations, and will be inferred from the original

jakemac53 opened the pull request:
Closes https://github.com/dart-lang/language/issues/3879

davidmorgan commented:
Thanks! I suggest to tweak the blocks starting "The signature of the constructor augmentation" and "The signature of the function augmentation" at the same time, as the sense of these has now changed: you are not required to repeat everything, it won't necessarily be possible to see the full signature.

lrhn commented on working/augmentation-libraries/feature-specification.md:
```diff
@@ -318,6 +318,18 @@ It is a compile-time error if:
+is inferred.
+
+This applies to return types, parameter types, and variable types, but not type
+parameters.
```
That is, types, not names. 
What about type parameter bounds? Would be nice to inherit those.

jakemac53 commented:
> Thanks! I suggest to tweak the blocks starting "The signature of the constructor augmentation" and "The signature of the function augmentation" at the same time, as the sense of these has now changed: you are not required to repeat everything, it won't necessarily be possible to see the full signature.

Updated

lrhn commented on working/augmentation-libraries/feature-specification.md:
```diff
@@ -318,6 +318,19 @@ It is a compile-time error if:
+
+When a type is omitted in an augmenting declaration, if the first declaration
+that introduced the member has a type specified in the ommitted types position,
+then that type is inherited.
```
> if the first declaration ...  has a type specified

Even if the original declaration didn't have a type specified, *it* would have had one inferred, inherited or defaulted. That's the type that gets used for the omitted type of the augmentetion.

jakemac53 replied:
Fixes here https://github.com/dart-lang/language/pull/3948

lrhn commented on working/augmentation-libraries/feature-specification.md:
```diff
@@ -474,7 +487,7 @@ The augmenting function does not have to pass the same arguments to
 at all.
 
-It is a compile-time error if:
+It is a compile-time error if, after type inheritance:
```
(Type inference can't change the type of the augmenting declaration, it inherits its types, but it can only do so after type inference has given types to its augmented declaration, which is why this is "after type inference".)

lrhn reviewed (commented):
Pedantic nits. :)

davidmorgan commented:
@jakemac53 oh, now I know why I couldn't comment properly ... GitHub doesn't let you comment on unchanged regions :(

The two regions with "...accomplish anything semantically. But it ensures that anyone reading the..." should probably be removed, they no longer make sense since you can omit information if you want to.

jakemac53 commented:
> The two regions with "...accomplish anything semantically. But it ensures that anyone reading the..." should probably be removed, they no longer make sense since you can omit information if you want to.

I left these in, while the types are omitted the parameter names are still required.
